lang/LangPrimSource/SC_Clock.h

```diff
@@ -0,0 +1,269 @@
+/*
+	SuperCollider real time audio synthesis system
+    Copyright (c) 2002 James McCartney. All rights reserved.
+	http://www.audiosynth.com
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
+*/
+
+#ifndef SC_CLOCK_INCLUDED
+#define SC_CLOCK_INCLUDED
+
+#include "PyrKernel.h"
+#include "PyrSched.h"
+#include "GC.h"
+#include "PyrPrimitive.h"
+#include "PyrSymbol.h"
+
+// heaps use an integer timestamp to ensure stable heap order
+struct PyrHeap:
+	PyrObjectHdr
+{
+	PyrSlot count;    // stability count
+	PyrSlot slots[0]; // slots
+};
+
+//if ABLETON LINK functionality is on, we must define some methods virtual
+#ifdef SC_ABLETON_LINK
+#define CLOCK_VIRTUAL_METHOD virtual
+#else
+#define CLOCK_VIRTUAL_METHOD
+#endif
+
+class TempoClock
+{
+public:
+	TempoClock(VMGlobals *inVMGlobals, PyrObject* inTempoClockObj,
+				double inTempo, double inBaseBeats, double inBaseSeconds);
+	CLOCK_VIRTUAL_METHOD ~TempoClock() {}
+	void StopReq();
+	void Stop();
+	void StopAndDelete()
+	{
+		Stop();
+		delete this;
+	}
+
+	void* Run();
+
+	void Add(double inBeats, PyrSlot* inTask);
+	void SetTempoAtBeat(double inTempo, double inBeats);
+	void SetTempoAtTime(double inTempo, double inSeconds);
+	void SetAll(double inTempo, double inBeats, double inSeconds);
+	double ElapsedBeats();
+	void Clear();
+	//void Flush();
+	double GetTempo() const { return mTempo; }
+	double GetBeatDur() const { return mBeatDur; }
+	CLOCK_VIRTUAL_METHOD double BeatsToSecs(double beats) const
+		{ return (beats - mBaseBeats) * mBeatDur + mBaseSeconds; }
+	CLOCK_VIRTUAL_METHOD double SecsToBeats(double secs) const
+		{ return (secs - mBaseSeconds) * mTempo + mBaseBeats; }
+	void Dump();
+
+//protected:
+	VMGlobals* g;
+	PyrObject* mTempoClockObj;
+	PyrHeap* mQueue;
+
+	double mTempo; // beats per second
+	double mBeatDur; // 1/tempo
+	double mBeats; // beats
+	double mBaseSeconds;
+	double mBaseBeats;
+	bool mRun;
+	SC_Thread mThread;
+	condition_variable_any mCondition;
+	TempoClock *mPrev, *mNext;
+
+	static TempoClock *sAll;
```

Comment:
`sAll`, `mNext`, `mPrev` and the `TempoClock_stopAll` function are obsolete because `TempoClock.all` is now handled completely in sclang. Is it appropriate to remove them in this branch ?

Reply:
Nevermind I just realised that `TempoClock_stopAll` is used during shutdown.